src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2018 Elfarouk Harb
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+
+ #include <shogun/optimization/FirstOrderSAGCostFunctionInterface.h>
+ using namespace shogun;
+
+ FirstOrderSAGCostFunctionInterface::FirstOrderSAGCostFunctionInterface(
+   Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* X,
+   Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y,
+   Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* trainable_parameters,
+   std::vector< std::function < stan::math::var(int32_t) > >* cost_for_ith_point,
+   std::function < stan::math::var(std::vector<stan::math::var>*) >* total_cost
+ )
+ {
+   m_X                    = X;
+   m_y                    = y;
+   m_trainable_parameters = trainable_parameters;
+   m_cost_for_ith_point   = cost_for_ith_point;
+   m_total_cost           = total_cost;
+ }
+
+ void FirstOrderSAGCostFunctionInterface::set_training_data(
+   Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic> * X_new,
+   Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y_new
+ )
+ {
+  REQUIRE(X_new!=NULL, "X must be non NULL");
```

Comment:
Is this a user facing error?
If no, pls say "No X provided.\n" (also missing newline)